bee-network/bee-autopeering/src/peering/manager.rs

```diff
@@ -0,0 +1,774 @@
+
+    let peer_req = request_mngr.write().new_peering_request(*peer_id, response_tx, local);
+
+    let msg_bytes = peer_req.to_protobuf().expect("error encoding peering request").to_vec();
+
+    server_tx
+        .send(OutgoingPacket {
+            msg_type: MessageType::PeeringRequest,
+            msg_bytes,
+            peer_addr,
+        })
+        .expect("error sending peering request to server");
+}
+
+/// Sends a peering response to a peer's address.
+pub(crate) fn send_peering_response_to_addr(
+    peer_addr: SocketAddr,
+    peer_id: &PeerId,
+    msg_bytes: &[u8],
+    tx: &ServerTx,
+    status: bool,
+) {
+    log::trace!("Sending peering response to: {}", peer_id);
+
+    let request_hash = msg_hash(MessageType::PeeringRequest, msg_bytes).to_vec();
+
+    let peer_res = PeeringResponse::new(request_hash, status);
+
+    let msg_bytes = peer_res
+        .to_protobuf()
+        .expect("error encoding peering response")
+        .to_vec();
+
+    tx.send(OutgoingPacket {
+        msg_type: MessageType::VerificationResponse,
+        msg_bytes,
+        peer_addr,
+    })
+    .expect("error sending peering response to server");
+}
+
+/// Sends a drop-peering request to a peer.
+pub(crate) fn send_drop_peering_request_to_peer(
+    peer: Peer,
+    server_tx: &ServerTx,
+    event_tx: &EventTx,
+    inbound_nbh: &InboundNeighborhood,
+    outbound_nbh: &OutboundNeighborhood,
+) {
+    let peer_port = peer
+        .port(AUTOPEERING_SERVICE_NAME)
+        .expect("missing autopeering service");
+    let peer_addr = SocketAddr::new(peer.ip_address(), peer_port);
+    let peer_id = peer.into_id();
+
+    send_drop_peering_request_to_addr(peer_addr, peer_id, server_tx, event_tx, inbound_nbh, outbound_nbh);
+}
+
+/// Sends a drop-peering request to a peer's address.
+pub(crate) fn send_drop_peering_request_to_addr(
+    peer_addr: SocketAddr,
+    peer_id: PeerId,
+    server_tx: &ServerTx,
+    event_tx: &EventTx,
+    inbound_nbh: &InboundNeighborhood,
+    outbound_nbh: &OutboundNeighborhood,
+) {
+    log::trace!("Sending drop request to: {}", peer_id);
+
+    let msg_bytes = DropPeeringRequest::new()
+        .to_protobuf()
+        .expect("error encoding drop request")
+        .to_vec();
+
+    server_tx
+        .send(OutgoingPacket {
+            msg_type: MessageType::DropRequest,
+            msg_bytes,
+            peer_addr,
+        })
+        .expect("error sending drop-peering request to server");
+
+    publish_drop_peering_event(peer_id, event_tx, inbound_nbh, outbound_nbh);
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////
+// EVENTS
+///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+/// Publishes the corresponding peering event [`IncomingPeering`], or [`OutgoingPeering`].
+pub(crate) fn publish_peering_event<const IS_INBOUND: bool>(
```

Comment:
I have seen this generic pretty much everywhere but I am not sure I understand the benefits of it. Are we really just trying to avoid a couple of runtime `if`s or is there more to it ? In the context of autopeering, these ifs sound really negligible...

Reply:
It's not related to that at all. I find it much cleaner this way and a great simple way of using const generics. The `InboundNeighborhood` and the `OutboundNeighborhood` are two distinct things, that share the same underlying datastructure, but are semantically are different things nonetheless. In Rust I like to have different types for semantically different things (Inbound/OutboundNeighborhood) instead of just a `Neighborhood`, which then differentiates on each API call whether we are in `inbound` or `outbound` state.

Reply:
If I wouldn't use const generics for this function here, I probably would have introduced two functions anyway:
`publish_inbound_peering_event` and `publish_outbound_peering_event`, which are both basically the same logic with things swapped out. That would make it more tedious to keep both always in sync. Or I'ld keep ONE function with branching at runtime: It have an annoying structure like:
```rust
if is_inbound {
    A;
    B;
    C;
else {
    A';
    B';
    C';
}
```